### ISSUE TITLE
`shell-pop-eshell` fails when treemacs buffer is open

### ISSUE THREAD

**Issue by TimoFreiberg** [Alexander-Miller/treemacs#17]
error message: `split-window: Cannot split side window or parent of side windowInvalid face reference: spacemacs-treemacs-face`

**Comment by Alexander-Miller** [comment]
works for me :(

SPC t D + stacktrace pls

**Comment by TimoFreiberg** [comment]
```
Debugger entered--Lisp error: (error "Cannot split side window or parent of side window")
  signal(error ("Cannot split side window or parent of side window"))
  error("Cannot split side window or parent of side window")
  split-window(#<window 11> 33 below)
  shell-pop-split-window()
  shell-pop-up(1)
  shell-pop(nil)
  (if (equal (quote (4)) index) (eshell nil) (shell-pop--set-shell-type (quote shell-pop-shell-type) (backquote ("eshell" "*eshell*" (lambda nil (eshell nil))))) (shell-pop index))
  spacemacs/shell-pop-eshell(nil)
  funcall-interactively(spacemacs/shell-pop-eshell nil)
  call-interactively(spacemacs/shell-pop-eshell nil nil)
  command-execute(spacemacs/shell-pop-eshell)
```

**Comment by ambihelical** [comment]
I can reproduce this shell-pop w/o spacemacs, but it only happens when treemacs has the cursor.  
But isn't this just a limitation of any side window though?

**Comment by Anntoin** [comment]
Also seeing this behaviour, same error. Was just trying the treemacs layer in the hopes this would work
For reference here's the neotree issue https://github.com/syl20bnr/spacemacs/issues/7446
Would be great if treemacs was able to fix this

**Comment by Alexander-Miller** [comment]
@ambihelical is right - it's a general side window limitation. I am currently busy making sure treemacs' side window status is kept intact when using eyebrowse and persp (though this issue is making me doubt whether it should have it in the first place). When this is done I should have the means I need to fix this issue.

In the meantime you can use eyebrowse as a workaround - switch to a different workspace and back and treemacs loses its side window status, which means that 1) the treemacs window gets closed by `delete-other-windows` or `magit-status` and 2) functions like pop-shell-* work again.

**Comment by Anntoin** [comment]
@Alexander-Miller thanks for the info, understanding the side window problem led me to another workaround. For me I just now  set `shell-default-full-span nil` so the terminal doesn't split the side window. For my usage this is perfect and works for treemacs and neotree.

**Comment by Alexander-Miller** [comment]
Treemacs is no longer a side window since v1.9.1. Popping shells and similar functions should work as usual now.

@TimoFreiberg You'll obviously be the guinea pig for this and the spacemacs layer after the exchange tomorrow ;)

**Comment by TimoFreiberg** [comment]
Just tested it, works!

**Comment by ambihelical** [comment]
I'm not really finding it to work very well.  In fact it's pretty weird.  This is using e1228815, master HEAD as of 15 Sept 2017:


      (1)       orig       (2)
     TTTT       TBBB       BBB
     tPPP   <=  TBBB  =>   TPP
     tPPP       TBBB       TPP


Center is the original configuration. 
Where T is a treemacs buffer opened on the left. B is some other buffer and P is pop-shell. 
I then try pop-shell with the cursor in (1) the treemacs buffer. (2) the other buffer. 

In (1) Treemacs splits into T and t.  The original buffer goes away. 
In (2) The shell goes to the bottom right, the original buffer is at top, and treemacs the bottom 2/3 of left. 

I can open another issue if anyone else is seeing the same.

**Comment by Alexander-Miller** [comment]
Let's keept going here.

It works for me, so I guess it's matter of configuration. Can you provide me with your values in `customize-group shell-pop`?

Also what do you use for window management - shackle, popwin, or purpose?

**Comment by ambihelical** [comment]
Here's my use-package config for shell-pop (I don't use customize, custom-file is "/dev/null"). 
I have played with shell-pop-full-span, neither t or nil "fixes" the problem, but I didn't keep track of the differences. 

``` 
(use-package shell-pop
  :config
  :general
    ("<f4> t"     #'shell-pop)
  :init
  (setq shell-pop-internal-mode "ansi-term"
        shell-pop-term-shell "/bin/bash"
        shell-pop-window-size 40
        shell-pop-window-position "top"
        shell-pop-full-span t
        shell-pop-universal-key "<f4> t"))

```

I use popwin.  Here is it's config:

```
(use-package popwin
  :init
  :config
  (push '("*Async Shell Command*" :noselect t) popwin:special-display-config)
  (push '("*Completions*" :stick t :noselect t) popwin:special-display-config)
  (push '("*Help*" :stick t ) popwin:special-display-config)
  (push '(Man-mode :stick t :height 20) popwin:special-display-config)
  (push '("*undo-tree*" :stick t :width 60 :position right) popwin:special-display-config)
  (push '("*General Keybindings*" :width 72 :position right) popwin:special-display-config)
  (popwin-mode 1)
  :defer 2)

```

I just tried disabling popwin via :disabled, but the result is unchanged.

**Comment by Alexander-Miller** [comment]
I've reproduced it now, though I needed to go full emacs -Q for it. Looks like something in spacemacs prevents the problem from happening.

On the other hand it looks like it's an issue with shell-pop and happens even when you just split the frame into two windows. Using your settings, and nothing else, I went from [this](https://i.imgur.com/uR65h0n.png) to [this](https://i.imgur.com/yRwKWEk.png).

I was just about to say you should open an issue againt shell-pop, but [one's open already](https://github.com/kyagi/shell-pop-el/issues/51). Judging by the comments telling popwin to open shell buffers in the current window should do the trick.

**Comment by ambihelical** [comment]
My comment didn't really address the original bug caused by it being a side window, just that the new implementation acted strangely.  I don't think there is any reason to keep this one open.

**Comment by et2010** [comment]
Thanks for the quick reply!

Here's what I got by running your snippet after each steps:

After step 3, I got:

```
((split-window . visual-fill-column-split-window) (quit-restore same (#<buffer session2.py> 1 #<marker at 5997 in session2.py> 137) #<window 10 on  *Treemacs-Framebuffer-1*> #<buffer  *Treemacs-Framebuffer-1*>) (linum--margin 4 . 4))
```
and after step 4 ( running magit status), I got

```
((window-side . left) (split-window . visual-fill-column-split-window) (quit-restore same (#<buffer session2.py> 1 #<marker at 5997 in session2.py> 137) #<window 10 on  *Treemacs-Framebuffer-1*> #<buffer  *Treemacs-Framebuffer-1*>) (linum--margin 4 . 4))
```

So it's indeed magit causing the problem.

**Comment by Alexander-Miller** [comment]
Before we go any further with magit make sure you're still running into this with the newest (and uncustomized) magit version.

**Comment by Alexander-Miller** [comment]
I may have found a clue to follow.
What's your value of `magit-display-buffer-function` and does the bug persist when you set it to  `'magit-display-buffer-fullframe-status-v1`?

**Comment by et2010** [comment]
Its value is `magit-display-buffer-traditional` and the bug didn't persist  when setting it to `magit-display-buffer-fullframe-status-v1`.

**Comment by Alexander-Miller** [comment]
It is a workaround, but not one that should be necessary in the first place. I've looked around the magit code and it looks like it's just using emacs window-config buitlins.

I suppose we need help straight from the source. Hello @tarsius, can you shed some light on magit's display/bury buffer internals seemingly restoring the wrong window parameters?

**Comment by tarsius** [comment]
See [Quitting Windows](https://magit.vc/manual/magit/Quitting-Windows.html) for an overview. Then you probably should also inspect the mentioned functions.

**Comment by Alexander-Miller** [comment]
Yes, those are the functions I went through and did not find anything suspicious. Since the issue appears with a default config I assume that the value of `magit-bury-buffer-function` is `magit-restore-window-configuration`. If so we need to do some deep digging. @et2010, here's what you need to do:

 * Find your magit installation and run `eval-buffer` in its .el files, so we'll have a readable stacktrace without bytecode.
 * `toggle-debug-on-error` and make sure it's enabled.
 * Eval the folowing code:

```
(defun debug-set-window-parameter (func &rest args)
  (message "[DEBUG] Set Window Parameter %s" args)
  (if (and (s-starts-with? " *Treemacs-Framebuffer" 
                           (-> args (car) (window-buffer) (buffer-name)))
           (eq 'window-side (cadr args)))
      (error "Setting window side param of treemacs!"))
  (apply func args))
(advice-add #'set-window-parameter :around #'debug-set-window-parameter)

```

 * verify it's working by running the following when treemacs is visible: `(set-window-parameter (treemacs--is-visible?) 'window-side 1)`
 * run your reproduction steps and post the resulting stacktrace.

**Comment by et2010** [comment]
Sorry for the delayed reply. I've followed your instructions and all went well except the magit status window didn't show up while treemacs was visible. I've not bumped into any errors so no trace back is provided, and the echo message when running magit status is `Setting window side param of treemacs!`

**Comment by Alexander-Miller** [comment]
> split-window-right

I think I know what this is about now. There was a bug, or rather incompatibility, in treemacs when spitting windows like this that made the resulting windows the wrong size. That is because the treemacs window does not really have a fixed width, it just instantly corrects itself whenever the window config changes. This was necessary because having a fixed width caused errors when e.g. calling `balance-windows` (this seems no longer be the case, maybe I can remove the workaround alltogether).

Now the solution to the split site issue was to treat treemacs as a side-window - but only for the duration of the split - since emacs' internals would then make the correct calculations. Unfortunately I did not correctly remove the side-window property again.

Check is the issue persists after you evaluate the following code (without the error throwing debug code from before in the way):
```
(defun treemacs--split-window-advice (original-split-function &rest args)
  "Advice to make sure window splits are sized correctly with treemacs.
This will treat the treemacs window as a side-window for the duration of the
split, calling the ORIGINAL-SPLIT-FUNCTION with its ARGS. This prevents the
calculations in `split-window-right' from outputting the wrong result for the
width of the new window when the treemacs window is visible."
  (-let [w (treemacs--is-visible?)]
    (unwind-protect
        (progn
          (when w (set-window-parameter w 'window-side treemacs-position))
          (apply original-split-function args))
      (when w (set-window-parameter w 'window-side nil)))))
```

**Comment by et2010** [comment]
Thanks! It works like a charm!
